woocommerce/woocommerce
# Add Marketplace license management endpoints

### Changes proposed in this Pull Request:

This is part of the [feature/marketplace-subscriptions](https://github.com/woocommerce/woocommerce/tree/feature/marketplace-subscriptions) project.

This PR implements three Marketplace API endpoints that will be used by the new My subscriptions page.

The `refresh` endpoint refreshes account and subscription data.
The `activate` endpoint activates a license.
The `deactivate` endpoint deactivates a license.



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Send a request to the refresh endpoint
```
curl --location --request POST 'http://localhost:8888/wp-json/wc/v3/marketplace/refresh'
```
2. Confirm it fails with a `rest_forbidden` error
3. Disable permission checks for testing by [returning true on this line](https://github.com/woocommerce/woocommerce/blob/cb3ca6080e616fc12e9a9c1b5c86d55fbe6d9c57/plugins/woocommerce/includes/admin/helper/class-wc-helper-subscriptions-api.php#L77)
4. Buy a subscription on WCCOM (can be a free product)
5. Send a request to the refresh endpoint
```
curl --location --request POST 'http://localhost:8888/wp-json/wc/v3/marketplace/refresh'
```
6. Confirm that the response includes your new subscription
7. Copy the `product_key` value for your new subscription
8. Send a request to the activation endpoint (replace <PRODUCT-KEY> with your key)
```
curl --location 'http://localhost:8888/wp-json/wc/v3/marketplace/subscriptions/activate' \
--form 'product_key="<PRODUCT-KEY>"'
```
9. Send a request to the subscriptions endpoint
```
curl --location 'http://localhost:8888/wp-json/wc/v3/marketplace/subscriptions'
```
10. Confirm that your subscription is now active
11. Send a request to the deactivation endpoint (replace <PRODUCT-KEY> with your key)
```
curl --location 'http://localhost:8888/wp-json/wc/v3/marketplace/subscriptions/deactivate' \
--form 'product_key="<PRODUCT-KEY>"'
```
12. Send a request to the subscriptions endpoint
```
curl --location 'http://localhost:8888/wp-json/wc/v3/marketplace/subscriptions'
```
13. Confirm that your subscription is now inactive



### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 
Add refresh, activate, and deactivate endpoints to the Marketplace API.

</details>
